src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs

```diff
@@ -14,148 +14,287 @@ internal struct ReadStack
+                    else
+                    {
+                        AddCurrent();
+                        _count--;
+                        _continuationCount = _count;
+                        _count--;
+                        Current = _previous[_count - 1];
+                    }
+
+                    return;
                 }
-                else
+
+                if (_continuationCount == 1)
                 {
-                    sb.Append('.');
-                    sb.Append(propertyName);
+                    // No need for a pop since there is only one stack frame.
+                    Debug.Assert(_count == 1);
+                    return;
                 }
+
+                // Update the list entry to the current value.
+                _previous[_count - 1] = Current;
+
+                Debug.Assert(_count > 0);
+            }
+            else
+            {
+                Debug.Assert(_continuationCount == 0);
+            }
+
+            if (_count > 1)
+            {
+                Current = _previous[--_count -1];
             }
         }
 
-        private string? GetPropertyName(in ReadStackFrame frame)
+        // Return a JSONPath using simple dot-notation when possible. When special characters are present, bracket-notation is used:
+        // $.x.y[0].z
+        // $['PropertyName.With.Special.Chars']
+        public string JsonPath()
         {
-            // Attempt to get the JSON property name from the frame.
-            byte[]? utf8PropertyName = frame.JsonPropertyName;
-            if (utf8PropertyName == null)
+            StringBuilder sb = new StringBuilder("$");
+
+            // If a continuation, always report back full stack.
+            int count = Math.Max(_count, _continuationCount);
+
+            for (int i = 0; i < count - 1; i++)
             {
-                // Attempt to get the JSON property name from the JsonPropertyInfo.
-                utf8PropertyName = frame.JsonPropertyInfo?.JsonPropertyName;
+                AppendStackFrame(sb, _previous[i]);
             }
 
-            string? propertyName;
-            if (utf8PropertyName != null)
+            if (_continuationCount == 0)
             {
-                propertyName = JsonHelpers.Utf8GetString(utf8PropertyName);
+                AppendStackFrame(sb, Current);
             }
-            else
+
+            return sb.ToString();
+
+            void AppendStackFrame(StringBuilder sb, in ReadStackFrame frame)
             {
-                propertyName = null;
+                // Append the property name.
+                string? propertyName = GetPropertyName(frame);
+                AppendPropertyName(sb, propertyName);
+
+                if (frame.JsonClassInfo != null)
+                {
+                    if (frame.IsProcessingEnumerable())
+                    {
+                        IEnumerable enumerable = (IEnumerable)frame.ReturnValue!;
+                        if (enumerable != null)
```

Comment:
Not sure what you mean.

Reply:
Trying to reduce the level of if-branch nesting to 2-3.

Something like this:
```C#
static void AppendStackFrame(StringBuilder sb, in ReadStackFrame frame)
{
    // Append the property name.
    string? propertyName = GetPropertyName(frame);
    AppendPropertyName(sb, propertyName);

    if (frame.JsonClassInfo != null)
    {
        if (frame.IsProcessingEnumerable())
        {
            IEnumerable? enumerable = (IEnumerable?)frame.ReturnValue;
            if (enumerable == null)
            {
                return;
            }

            // Once all elements are read, the exception is not within the array.
            if (frame.ObjectState < StackFrameObjectState.ReadElements)
            {
                sb.Append(@"[");
                sb.Append(GetCount(enumerable));
                sb.Append(@"]");
            }
        }
    }
}
```

Or even better, merge the conditions to have less if-nesting like this:
```C#
static void AppendStackFrame(StringBuilder sb, in ReadStackFrame frame)
{
    // Append the property name.
    string? propertyName = GetPropertyName(frame);
    AppendPropertyName(sb, propertyName);

    if (frame.JsonClassInfo != null && frame.IsProcessingEnumerable())
    {
        IEnumerable? enumerable = (IEnumerable?)frame.ReturnValue;
        if (enumerable == null)
        {
            return;
        }

        // Once all elements are read, the exception is not within the array.
        if (frame.ObjectState < StackFrameObjectState.ReadElements)
        {
            sb.Append(@"[");
            sb.Append(GetCount(enumerable));
            sb.Append(@"]");
        }
    }
}
```

Reply:
I see. Yes it's a trade-off to whether adding a `return` to a method without one is better or not.

Also a trade-off for debugging via breakpoints for adding the `&&` clause which doesn't allow an easy way to set a breakpoint for the first condition. Since I'm done debugging here I may as well collapse that into a single `if`.